### DISCUSSION TITLE
Installed package via -pip only works as superuser (docker)

### DISCUSSION THREAD

**Discussion by smelczer** [sagemathinc/cocalc#6873] [Q&A] [answered]
I'm been using the (fantastic) cocalc docker container recently and ran into a strange bug: when installing an external package it works perfectly under `sudo sage` but does not work without superuser privileges. My process was the following 

(1) Install cocalc docker on Ubuntu server using 
```
apt install docker.io
docker run --name=cocalc -d -v ~/cocalc:/projects -p 443:443 sagemathinc/cocalc-v2
```

(2) When docker setup is complete install the ore_algebra package with
```
sudo docker exec -it cocalc bash
sage -pip install git+https://github.com/mkauers/ore_algebra.git
```

(3) Go into the cocalc server and open a terminal, then call

```
sage
from ore_algebra import *  # No issues, seems like the package is found
OreAlgebra # One of the functions in the package
```

This gives `NameError: name 'OreAlgebra' is not defined`. The same behaviour happens in a Jupyter notebook. However, giving the project sudo access and running

```
sudo sage
from ore_algebra import *  
OreAlgebra
```

finds the OreAlgebra function and behaves correctly!  

Any ideas about what could be going wrong? I don't know of another non-default sage package that can be installed with pip to see if this is specific to the `ore_algebra` package, or holds for other external packages installed with pip.

**Comment by williamstein** [comment, marked as answer]
You need to do

umask 022

first or the permissions of the files created by root will be 007 so normal
users can't read them.

On Wed, Aug 16, 2023 at 11:08 PM Stephen Melczer ***@***.***>
wrote:

> I'm been using the (fantastic) cocalc docker container recently and ran
> into a strange bug: when installing an external package it works perfectly
> under sudo sage but does not work without superuser privileges. My
> process was the following
>
> (1) Install cocalc docker on Ubuntu server using
>
> apt install docker.io
> docker run --name=cocalc -d -v ~/cocalc:/projects -p 443:443 sagemathinc/cocalc-v2
>
> (2) When docker setup is complete install the ore_algebra package with
>
> sudo docker exec -it cocalc bash
> sage -pip install git+https://github.com/mkauers/ore_algebra.git
>
> (3) Go into the cocalc server and open a terminal, then call
>
> sage
> from ore_algebra import *  # No issues, seems like the package is found
> OreAlgebra # One of the functions in the package
>
> This gives NameError: name 'OreAlgebra' is not defined. The same
> behaviour happens in a Jupyter notebook. However, giving the project sudo
> access and running
>
> sudo sage
> from ore_algebra import *
> OreAlgebra
>
> finds the OreAlgebra function and behaves correctly!
>
> Any ideas about what could be going wrong? I don't know of another
> non-default sage package that can be installed with pip to see if this is
> specific to the ore_algebra package, or holds for other external packages
> installed with pip.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sagemathinc/cocalc/discussions/6873>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJXS5V4B2AKJKNXA5NBJYLXVWYNFANCNFSM6AAAAAA3TRTLCM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
-- William Stein

**Reply by smelczer** [reply]
Fixes the problem, thanks

**Reply by williamstein** [reply]
Any ideas about how I could document this better?

**Reply by smelczer** [reply]
It is prominently displayed on the GitHub page, and should have been clear. I was flipping back and forth between multiple sources to get everything set up and somehow missed it. 

On a related note, I documented my process of setting up the docker container on an nginx server, serving from a custom domain, and getting an ssl certificate from letsencrypt. If you think this information would be useful to others I could make a post about it on the cocalc-docker discussion forum (I don't have experience with nginx and related topics, but was able to cobble it together from the helpful sections on the git readme and some external tutorials -- this means it might not be 100% best practice, but could at least serve as a basis for others with little experience).

**Reply by williamstein** [reply]
Yes, please share that process regarding the cert!

**Reply by smelczer** [reply]
I posted the process I followed in the cocalc-docker discussions.

**Reply by williamstein** [reply]
Thanks!

**Reply by williamstein** [reply]
https://github.com/sagemathinc/cocalc-docker/discussions/200

**Reply by williamstein** [reply]
I also linked to this from the instructions section of the readme: https://github.com/sagemathinc/cocalc-docker/blob/master/README.md#instructions
